MNT: add set_low_lim and set_hi_lim to EpicsMotor

cristinasewell opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
## Description
- Added two methods to the `epics_motors`: `set_high_limit()` and `set_low_limit()`, as well as `get_low_limit()` and `get_high_limit()`.
- Added a little method to clear limits: `clear_limits` - this sets both limits to 0.
- Removed the other limit setters and getters `low_limit`, `high_limit` `limits`
- Overrode the `limits` property

<!--- Describe your changes in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Trying to solve the problem described in #597
## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Locally only.
## Where Has This Been Documented?
<!--  Include where the changes made have been documented. -->
<!--  This can simply be  a comment in the code or updating a docstring -->
Docstring
<!--
## Screenshots (if appropriate):
-->

## Pre-merge checklist
- [ ] Code works interactively
- [ ] Code contains descriptive docstrings, including context and API
- [ ] New/changed functions and methods are covered in the test suite where possible
- [ ] Test suite passes locally
- [ ] Test suite passes on travis
- [ ] Ran docs/pre-release-notes.sh and created a pre-release documentation page
- [ ] Pre-release docs include context, functional descriptions, and contributors as appropriate

cristinasewell commented on pcdsdevices/epics_motor.py:
```diff
@@ -221,6 +221,48 @@ def check_limit_switches(self):
+        value : float
+            Limit of travel in the negative direction.
+        """
+        if not self.moving:
```
Not sure if I need to check if if it is moving or not and if `<=` current position - this is just something the `ophyd` guys are using in `set_lim()`

ZLLentz replied:
Ah that's interesting- I wonder what kinds of problems can happen from mid-move limit sets...
I'm a bit surprised that ophyd fails silently instead of raising an exception

klauer replied:
It's a recent addition - not entirely sure why it's there. I'm positive I've set limits when a motor was moving without issue. Perhaps just a race if you set a limit very close to the current position and the readback changes beyond that value? 🤷‍♂️

ZLLentz commented:
I think adding these two functions was a good idea.

As for the limit attributes... I'm not really sure what to do with them. I'm not 100% sure we want the properties at all- I'd like more opinions on this.

My point of view is that properties that simply get EPICS values slows down attribute inspection tools and can cause things to fail in unexpected ways without adding much utility.

I think if we just removed them entirely, the parent class's `limits` attribute is fully sufficient.

ZLLentz commented on pcdsdevices/epics_motor.py:
```diff
@@ -221,6 +221,48 @@ def check_limit_switches(self):
+            if value <= self.position:
+                self.low_limit_travel.put(value)
+            else:
+                logger.debug('Could not set motor low limit to %f at '
```
ophyd devices typically use `self.log` for this sort of thing- I think we still have a bunch of dangling `logger` references elsewhere in the code in pcdsdevices

cristinasewell replied:
ohh thank you, i'll change this

cristinasewell replied:
fixed this in 3290db5

cristinasewell commented:
> I think adding these two functions was a good idea.
> 
> As for the limit attributes... I'm not really sure what to do with them. I'm not 100% sure we want the properties at all- I'd like more opinions on this.
> 
> My point of view is that properties that simply get EPICS values slows down attribute inspection tools and can cause things to fail in unexpected ways without adding much utility.
> 
> I think if we just removed them entirely, the parent class's `limits` attribute is fully sufficient.

I removed them in 3290db5 - but I am not sure if did not break something else by doing that.

ZLLentz commented:
I still want more opinions on this- but I think the current code is a very good scenario:
- There is one canonical way to check the limits via getattr, the `limits` attribute, and it runs without any EPICS actions
- There are the following ways to set the limits:
  - ophyd's built-in `set_lim` (sets both, but you need to have both values ready)
  - the new `set_low_lim` (does checks and puts to the signal)
  - the new `set_hi_lim` (does checks and puts to the signal)
- There are additional `get_low_lim` and `get_hi_lim` functions that, while not strictly necessary, are easy to find if you know about the set variants.

If we have non-EPICS motors we can follow the same API with little issue.

My only suggestions are:
- Failing to set the limits for any reason should be an exception raised or a log message, not a silent failure. I think ophyd does it wrong here in `set_lim`.
- There might need to be some handling for if the user puts a high limit that's lower than the low limit, or a low limit that's higher than the high limit
- Do we need a shortcut for removing the soft limits entirely? (Setting high and low to zero at the same time)

klauer commented:
Some thoughts:
* I guess the `_lim` naming is following ophyd's `set_lim` - I don't particularly like the naming but that's OK.
* Completely agreed with Zach: this needs to raise in all conditions in which this fails silently (moving or position outside of limit).
* Limits do need special handling when `requested_low_limit > current_high_limit` (and similar scenarios). Try it with a motor record to see for yourself.
* `clear_limits` seems a nice shortcut as Zach indicated

cristinasewell commented:
> Some thoughts:
> 
> * I guess the `_lim` naming is following ophyd's `set_lim` - I don't particularly like the naming but that's OK.
> * Completely agreed with Zach: this needs to raise in all conditions in which this fails silently (moving or position outside of limit).
> * Limits do need special handling when `requested_low_limit > current_high_limit` (and similar scenarios). Try it with a motor record to see for yourself.
> * `clear_limits` seems a nice shortcut as Zach indicated

Do you think I should change it to something like this: `set_low_limit` and `set_high_limit` instead @klauer ? or what exactly you don't like about the naming?

cristinasewell commented on pcdsdevices/epics_motor.py:
```diff
@@ -191,7 +155,8 @@ def check_value(self, value):
         # command will be accepted by the motor
         if any(self.limits):
-            if not (self.low_limit <= value <= self.high_limit):
+            if not (self._get_epics_limits()[0] <= value
```
I was not sure here if I should have changed this. The `low_limit` and `high_limit` work as well (at least in some situations), because these are inherited from the ophyd Component (I think?)

ZLLentz replied:
For reference on inherited `low_limit` and `high_limit` attributes: https://github.com/bluesky/ophyd/blob/c91317fb21506d6acc2900e392e667aeb2684abc/ophyd/positioner.py#L142-L148

Is the concern here just the case where limits are uninitialized or invalid? It may be easiest to just wrap the `_get_epics_limits` check into an override of the `limits` attribute in that case. Then all instances of `self.limits`, `self.low_limit`, `self.high_limit` work as expected.

cristinasewell replied:
Overrode the `limits` attribute in 175f3e1 - and changed to using `limits` in 44f9852

cristinasewell commented on pcdsdevices/epics_motor.py:
```diff
@@ -221,6 +186,85 @@ def check_limit_switches(self):
+                if value < _current_high_limit:
+                    self.low_limit_travel.put(value)
+                else:
+                    raise ValueError('Could not set motor low limit to %f. '
```
This might be a bit too much, I probably could have combined the if statements here but I though raising more specific errors is not a bad idea?

klauer replied:
I missed this when looking earlier. Descriptive errors are very helpful 👍

cristinasewell commented on pcdsdevices/epics_motor.py:
```diff
@@ -221,6 +186,85 @@ def check_limit_switches(self):
+        """
+        if not self.moving:
+            # update EPICS
+            _current_low_limit = self._get_epics_limits()[0]
```
I'm using the `_get_epics_limits` here, because it will return `0` if not set - is this fine?

klauer commented on pcdsdevices/epics_motor.py:
```diff
@@ -221,6 +186,85 @@ def check_limit_switches(self):
         else:
             return "Low [] High []"
 
+    def get_low_limit(self):
```
I prefer the new names for sure - but you didn't _have_ to change them :)
Please also update the tab whitelist with the new names.

cristinasewell replied:
updated them in b3a7c93 - thanks for pointing this out

klauer commented on pcdsdevices/epics_motor.py:
```diff
@@ -221,6 +186,85 @@ def check_limit_switches(self):
+                                 'position %f. Low limit must be lower than '
+                                 'the current position.', value, self.position)
+        else:
+            raise ValueError('Could not set the low limit, the motor is in'
```
The control flow would be much more obvious if you inverted the `if not self.moving:` and bailed early.
That is, 
```python
if self.moving:
   raise ValueError('...')
# otherwise, do this:
```

cristinasewell replied:
Thanks - i have changed this in b3a7c93

klauer replied:
Nice - so much cleaner, don't you think? 👍

cristinasewell replied:
Yes of course, thank you 😊

klauer commented on pcdsdevices/epics_motor.py:
```diff
@@ -221,6 +186,85 @@ def check_limit_switches(self):
+                                     'current high limit: %f',
+                                     value, _current_high_limit)
+            else:
+                raise ValueError('Could not set motor low limit to %f at '
```
Careful - uninterpolated `%f` in your `ValueError`. Should use f-strings here.

cristinasewell replied:
Thank you, changed it to using f-strings in my ValueErrors.... b3a7c93

ZLLentz commented on pcdsdevices/epics_motor.py:
```diff
@@ -221,6 +186,80 @@ def check_limit_switches(self):
+    def clear_limits(self):
+        """Set both low and high limits to 0."""
+        self.set_low_limit(0)
+        self.set_high_limit(0)
```
This could fail if the limits are something like (-100, -50) before the function call, since the low limit can't be zero if the high limit is -50 due to the checks above. Funnily enough, `self.set_lim(0, 0)` also fails for similar reasons.

Might be easiest just to do straight puts to the PVs.

cristinasewell replied:
Ok i changed it in 175f3e1 using the ` self.high_limit_travel.put(0)` and `self.low_limit_travel.put(0)` - i think that's what you meant. 

Thank you

klauer commented on pcdsdevices/epics_motor.py:
```diff
@@ -52,20 +51,18 @@ class EpicsMotorInterface(FltMvInterface, EpicsMotor):
+                     "disable", "check_limit_switches", "get_low_limit",
+                     "set_low_limit", "get_high_limit", "set_high_limit"]
 
     def __init__(self, *args, **kwargs):
```
Looks like `__init__` can be removed here?

cristinasewell replied:
Did we only have this because of the `_limits`?

cristinasewell replied:
took it out in 54b6349

klauer reviewed (approved):
This looks good and ready to go to me. 

User experience when testing this may require some changes down the line, but we'll deal with that when we get there.

cristinasewell commented on docs/source/upcoming_release_notes/597-Motor_Soft_Limits_Consistency.rst:
```diff
@@ -0,0 +1,33 @@
+--------
+- Added two methods to the `epics_motors`: `set_high_limit()` and `set_low_limit()`, as well as `get_low_limit()` and `get_high_limit()`.
+- Added a little method to clear limits: `clear_limits` - this sets both limits to 0.
+- Removed the other limit setters and getters `low_limit`, `high_limit` `limits`
```
Ok thank you, i'll add this change specification to the `API Changes` section then and remove this last line from here

cristinasewell replied:
updated this a bit in 54b6349

ZLLentz commented:
I think this looks clean, I'm checking the merge conflict
